igcollect/strongswan.py

```diff
@@ -4,48 +4,86 @@
 Copyright (c) 2020 InnoGames GmbH
 """
 
+import binascii
+import re
+import vici
+
 from argparse import ArgumentParser
-from subprocess import check_output
+from hashlib import pbkdf2_hmac
 from time import time
-import re
+
+# Keep metric names consistent with what we have on other systems which probably
+# comes from SNMP. Translate from metrics used by Strongswan.
+METRICS = {
+    'bytesIn': 'bytes-in',
+    'bytesOut': 'bytes-out',
+    'pktsIn': 'packets-in',
+    'pktsOut': 'packets-out',
+}
 
 
 def parse_args():
     parser = ArgumentParser()
-    parser.add_argument('--prefix', default='strongswan')
+    parser.add_argument(
+        '--prefix', help='Graphite path prefix', type=str, default='strongswan',
+    )
+    parser.add_argument(
+        '--salt', help='Hash client names using this salt when it\'s specified',
+        type=str,
+    )
     return parser.parse_args()
 
 
 def main():
     args = parse_args()
+
+    # Get all connection and materialize enumerator into a list for later use.
+    sas = list(vici.Session().list_sas())
+
     template = args.prefix + '.{} {} ' + str(int(time()))
-    print(template.format('clients', count_connected_clients()))
-
-
-def count_connected_clients():
-    # Single connected users looks like this:
-
-    # ---- 8< ----
-    # vpn_by_id_group_xyz: #974, ESTABLISHED, IKEv2, 4242424242424242_i 4242424242424242_r*
-    #   local  'XXXXXX.example.com' @ XXX.XXX.XXX.XXX[4500]
-    #   remote 'XXX.XXX.XXX.XXX' @ XXX.XXX.XXX.XXX[56128] EAP: 'user.name' [XXX.XXX.XXX.XXX XXX:XXX:XXX::XXX]
-    #   AES_CBC-256/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
-    #   established 6648s ago, rekeying in 6393s
-    #   vpn_by_id_group_xyz: #2624, reqid 426, INSTALLED, TUNNEL-in-UDP, ESP:AES_CBC-256/HMAC_SHA1_96
-    #     installed 3354s ago, rekeying in 53s, expires in 606s
-    #     in  42424242 (-|0x000675ec), 9577036 bytes, 63405 packets,     2s ago
-    #     out 42424242 (-|0x000675ec), 20466058 bytes, 68523 packets,     2s ago
-    #     local  0.0.0.0/0 ::/0
-    #     remote XXX.XXX.XXX.XXX/32 XXX:XXX:XXX::XXX/128
-    # ---- >8 ----
-
-    # Let's say that the line containing word 'remote' and IP address of the client
-    # identifies a single client.
-    line_re = re.compile(r'^\s+remote .* @')
-
-    output = check_output(['/usr/sbin/swanctl', '--list-sas']).decode()
-    lines = output.splitlines()
-    return sum(1 for l in lines if line_re.match(l))
+    print(template.format('clients', len(sas)))
+
+    for client_name, client_data in get_clients_traffic(sas).items():
+        # For privacy we support hashing client login name.
+        if args.salt:
+            client_hash = binascii.hexlify(
```

Comment:
@raimund-s ?

Reply:
Using hashlib would be ok but we need to add a secret to the name as we want to directly dump these into our graphite. We need the paths to be stable per user but we don't want to make the usernames public to everyone who has access to our graphite. As an attacker could just easily test all possible user name and check which hash output matches this would be a case of a to small input data set for a hash to be irreversible.

Reply:
That should be possible. I would be strongly in favour to keep dependencies to third party modules to a minimum and use hashlib.

Reply:
There are no 3rd party modules used here.